helix-term/src/application.rs

```diff
@@ -53,33 +53,97 @@ pub struct Application {
+        let local_config_dir = helix_core::local_config_dir();
+        let cwd = std::env::current_dir().expect("unable to determine current directory");
+        if !config_dir.exists() {
+            std::fs::create_dir_all(&config_dir).ok();
+        }
 
-        let theme_loader =
-            std::sync::Arc::new(theme::Loader::new(&conf_dir, &helix_core::runtime_dir()));
-
-        // load default and user config, and merge both
-        let builtin_err_msg =
-            "Could not parse built-in languages.toml, something must be very wrong";
-        let def_lang_conf: toml::Value =
-            toml::from_slice(include_bytes!("../../languages.toml")).expect(builtin_err_msg);
-        let def_syn_loader_conf: helix_core::syntax::Configuration =
-            def_lang_conf.clone().try_into().expect(builtin_err_msg);
-        let user_lang_conf = std::fs::read(conf_dir.join("languages.toml"))
-            .ok()
-            .map(|raw| toml::from_slice(&raw));
-        let lang_conf = match user_lang_conf {
-            Some(Ok(value)) => Ok(merge_toml_values(def_lang_conf, value)),
-            Some(err @ Err(_)) => err,
-            None => Ok(def_lang_conf),
+        // `config.toml`
+        let mut config = match std::fs::read_to_string(config_dir.join("config.toml")) {
+            Ok(config) => toml::from_str(&config)
+                .map(crate::keymap::merge_keys)
+                .unwrap_or_else(|err| {
+                    eprintln!("Bad config: {}", err);
+                    eprintln!("Press <ENTER> to continue with default config");
+                    use std::io::Read;
+                    // This waits for an enter press.
+                    let _ = std::io::stdin().read(&mut []);
+                    Config::default()
+                }),
+            Err(err) if err.kind() == std::io::ErrorKind::NotFound => Config::default(),
+            Err(err) => return Err(Error::new(err)),
         };
 
+        const LANG_CONFIG_FILENAME: &str = "languages.toml";
+
+        // Config override order: local (cwd) -> local (root) -> global -> default.
+        // Read and parse the `languages.toml` files as TOML objects.
+        let default_lang_config: toml::Value =
+            toml::from_slice(include_bytes!("../../languages.toml")).unwrap_or_else(|_| {
+                panic!("failed to read the default `{}`", LANG_CONFIG_FILENAME)
+            });
+        let lang_config: toml::Value = {
+            let local_cwd_config = load_lang_config(cwd.join(".helix").join(LANG_CONFIG_FILENAME))?;
+            let local_root_config = load_lang_config(local_config_dir.join(LANG_CONFIG_FILENAME))?;
+            let global_config = load_lang_config(config_dir.join(LANG_CONFIG_FILENAME))?;
+            [
+                local_root_config,
+                global_config,
+                default_lang_config.clone(),
+            ]
+            .into_iter()
+            .fold(local_cwd_config, |a, b| merge_toml_values(b, a))
```

Comment:
Why not iterate over paths rather than lang config? That way we don't have to `load_lang_config` so many times.